### PR TITLE
fix(stage): derive approval order from tarballs

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

Batch `pnpm stage approve` now downloads every selected tarball before sending any approval request, reads each published `package.json`, and builds the dependency graph from those exact manifests. It approves selected dependencies first and skips dependents when a dependency cannot be approved.

The preflight bounds compressed and decompressed tarballs at 512 MiB, preserves npm aliases to tags, and rejects ambiguous batches containing two stages for the same package version before approving anything.

The behavior is implemented in both the TypeScript CLI and the Rust pnpm port. Tests cover preflight request ordering, dependency ordering, failure propagation, registry metadata mismatches, npm aliases, response limits, and duplicate package versions. The release documentation and existing changeset are updated.

## Squash Commit Body

```text
Batch stage approval derived its dependency order from the local workspace, which could differ from the selected staged artifacts or be unavailable on a maintainer machine.

Download and inspect every selected tarball before approving the batch. Build the graph from the published manifests, key relationships by stage ID, approve dependencies first, and propagate failed approvals to their selected dependents. Bound compressed and decompressed tarball reads, preserve alias-to-tag semantics, and reject duplicate staged identities before approval. Keep the TypeScript and Rust implementations aligned and share each implementation tarball reader with stage download.
```

## Checklist

- [x] The change is implemented in both the TypeScript CLI and the Rust
  `pnpm/` port, or the description notes what still needs porting.
- [x] Added a changeset (`pnpm changeset`) if this PR changes any published
  package. Keep it short and written for pnpm users. It becomes a release note.
- [x] Added or updated tests.
- [x] Updated the documentation if needed.

---
Written by an agent (Codex, GPT-5).

## Summary by CodeRabbit

* **New Features**
  * `stage approve` determines approval order from published manifests in selected tarballs, including dependencies outside the current workspace.
  * Selected tarballs are downloaded before approval, with dependency aliases and additional dependency types recognized.
  * Approval stops when selected dependencies are not approved.
  * Batches containing duplicate package versions are rejected before approval begins.

* **Bug Fixes**
  * Oversized tarballs are rejected safely during processing.

* **Documentation**
  * Updated release guidance to describe tarball-based dependency ordering.

### CODE REVIEW

File: pnpm11/releasing/commands/src/stage/tarball.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import { stageRequest } from './request.js'
+
+/** Download one staged package without writing it to disk. */
+export async function fetchStageTarball (context: StageContext, stageId: string): Promise<Buffer> {
+  const response = await stageRequest(context, {
+    url: new URL(`-/stage/${stageId}/tarball`, context.registry).href,
+    init: { method: 'GET' },
+    action: `download staged package ${stageId}`,
+  })
+  return Buffer.from(await response.arrayBuffer())
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm11/releasing/commands/src/stage/tarball.ts]: Fixed in 917e7c59f9. The TypeScript path now caps the streamed response at 512 MiB and cancels it on overflow. Gzip decompression uses the same maximum output size, and focused tests cover accepted, rejected, and cancelled response bodies.

---
Written by an agent (Codex, GPT-5).

File: pnpm11/releasing/commands/src/stage/approvalOrder.ts [hunk 2]
```diff
@@ -1,86 +1,113 @@
...
+    const parsed = npa.resolve(name, spec, process.cwd())
+    const registrySpec = parsed.type === 'alias' ? parsed.subSpec : parsed
+    if (
+      registrySpec?.name &&
+      (registrySpec.type === 'version' || registrySpec.type === 'range') &&
+      typeof registrySpec.fetchSpec === 'string'
+    ) {
+      return { name: registrySpec.name, spec: registrySpec.fetchSpec }
+    }
+  } catch {}
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm11/releasing/commands/src/stage/approvalOrder.ts]: Fixed in 917e7c59f9. The empty catch and parser call are gone. Alias parsing is now deterministic, and only npm aliases to a concrete semver version or range are normalized; aliases to tags retain their original package key in both implementations.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/cli/src/cli_args/stage/approve.rs [hunk 3]
```diff
@@ -355,103 +347,105 @@ fn is_missing_stage_error(error: &miette::Report) -> bool {
...
+fn manifest_for_graph(mut manifest: Value) -> Value {
+    for field in ["peerDependencies", "devDependencies", "optionalDependencies", "dependencies"] {
+        let Some(dependencies) = manifest.get(field).and_then(Value::as_object) else {
+            continue;
+        };
+        let normalized = dependencies
+            .iter()
+            .filter_map(|(name, spec)| {
+                let spec = spec.as_str()?;
+                let (name, spec) = PackageManifest::resolve_registry_dependency(name, spec);
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/cli_args/stage/approve.rs]: Fixed in 917e7c59f9. TypeScript and Rust now both normalize npm aliases only when the target is a version or range and preserve aliases to tags. Unit and integration coverage verifies the matching behavior and approval order in both stacks.

---
Written by an agent (Codex, GPT-5).

File: pnpm11/releasing/commands/src/tarball/readResponseBodyCapped.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    if (done) break
+    total += value.byteLength
+    if (total > maxBytes) {
+      // eslint-disable-next-line no-await-in-loop
+      await reader.cancel().catch(() => {})
+      return undefined
+    }
+    chunks.push(Buffer.from(value))
+  }
+  return Buffer.concat(chunks)
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm11/releasing/commands/src/tarball/readResponseBodyCapped.ts]: Fixed in 31a1ff7ea6. The capped reader now streams accepted chunks to a temporary file and reads one final buffer after closing it, so it no longer retains a full chunk set while allocating another full buffer. The multi-chunk regression test exercises the spooled path.

---
Written by an agent (Codex, GPT-5).

File: pnpm/crates/cli/src/cli_args/stage/approve.rs [hunk 5]
```diff
@@ -355,103 +348,136 @@ fn is_missing_stage_error(error: &miette::Report) -> bool {
...
+            continue;
+        };
+        let normalized = dependencies
+            .iter()
+            .filter_map(|(name, spec)| {
+                let spec = spec.as_str()?;
+                let (registry_name, registry_spec) =
+                    PackageManifest::resolve_registry_dependency(name, spec);
+                let (name, spec) = if Version::parse(registry_spec).is_ok()
+                    || Range::parse(registry_spec).is_ok()
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/cli/src/cli_args/stage/approve.rs]: Fixed in 56922a096c. Rust now maps a JavaScript-semver range containing an empty comparator set to `*`, matching the TypeScript graph semantics. A unit test covers normalization and the end-to-end metadata-mismatch test uses an out-of-range selected dependency to prove the edge is retained and approved first.

---
Written by an agent (Codex, GPT-5).

**Comment by zkochan** [comment]
Addressed the latest aggregate review findings in commits `d3c3faebd9` and `1dd647b3c4`:

- reused the existing npm and JavaScript-compatible semver parsers
- created response spool files exclusively with private permissions
- made duplicate tar manifests last-entry-wins in both stacks
- validated tarball package names and versions before graph construction
- added explicit preflight-order and malformed-manifest coverage
- aligned UTF-8 BOM handling across TypeScript and Rust

The two remaining Qodo notes are deliberately unchanged. Tarball preflight is serialized because each accepted tarball may be 512 MiB; bounded concurrency would multiply that peak memory, while approval safety requires every preflight to finish first. TypeScript function parameter commas are left in the form enforced by this repository’s Standard Style formatter and lint suite.

---
Written by an agent (Codex, GPT-5).

File: pnpm11/releasing/commands/src/stage/approvalOrder.ts [hunk 6]
```diff
@@ -1,86 +1,122 @@
...
-function orderIndexOf (item: ApprovalItem, order?: WorkspaceApprovalOrder): number {
-  const orderIndex = item.packageName ? order?.orderIndexByPackageName.get(item.packageName) : undefined
-  return orderIndex ?? Number.MAX_SAFE_INTEGER
+function parseRegistryDependency (name: string, spec: string): { name: string, spec: string } {
+  if (!spec.startsWith('npm:')) return { name, spec }
+  const parsed = npa.resolve(name, spec, process.cwd())
+  return (parsed.type === 'version' || parsed.type === 'range') &&
+    typeof parsed.name === 'string' &&
+    typeof parsed.fetchSpec === 'string'
+    ? { name: parsed.name, spec: parsed.fetchSpec }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm11/releasing/commands/src/stage/approvalOrder.ts:120]: This finding assumes the upstream `npm-package-arg` API. This repository pins the `@pnpm/npm-package-arg` 2.0.0 fork, whose `resolve()` implementation returns the parsed target itself via `Object.assign(npa(target), { alias, raw, rawSpec })`; it has the target `name`, `type`, and `fetchSpec`, and no `subSpec`. I also verified the installed package at runtime: `npm:@pnpmtest/stage-dependency@1.0.0` resolves to `type: version`, `name: @pnpmtest/stage-dependency`, and `fetchSpec: 1.0.0`. The focused test "stage approve derives package identity from the tarball instead of registry metadata" exercises this renamed scoped alias and asserts dependency-first approval. No code change is needed. Current head: `1dd647b3c4`.

---
Written by an agent (Codex, GPT-5).

File: pnpm11/releasing/commands/src/stage/approvalOrder.ts [hunk 7]
```diff
@@ -1,86 +1,122 @@
...
+    : { name, spec }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm11/releasing/commands/src/stage/approvalOrder.ts:121]: This finding assumes the upstream `npm-package-arg` API. This repository pins the `@pnpm/npm-package-arg` 2.0.0 fork, whose `resolve()` implementation returns the parsed target itself via `Object.assign(npa(target), { alias, raw, rawSpec })`; it has the target `name`, `type`, and `fetchSpec`, and no `subSpec`. I also verified the installed package at runtime: `npm:@pnpmtest/stage-dependency@1.0.0` resolves to `type: version`, `name: @pnpmtest/stage-dependency`, and `fetchSpec: 1.0.0`. The focused test "stage approve derives package identity from the tarball instead of registry metadata" exercises this renamed scoped alias and asserts dependency-first approval. No code change is needed. Current head: `1dd647b3c4`.

---
Written by an agent (Codex, GPT-5).

**Comment by zkochan** [comment]
@coderabbitai review

The only open finding from the previous review was resolved with source and runtime evidence for pnpm’s forked parser API. Please refresh the review state for the current head.

---
Written by an agent (Codex, GPT-5).
